File: trs/parse.ts
import { SExpr, Term, TRS } from "./types";

export function tokenize(input: string): string[] {
  const tokens: string[] = [];
  let i = 0;
  while (i < input.length) {
    if (/\s/.test(input[i])) { i++; continue; }
    if (input[i] === ";") {
      while (i < input.length && input[i] !== "\n") i++;
      continue;
    }
    if (input[i] === "(" || input[i] === ")") {
      tokens.push(input[i]);
      i++;
      continue;
    }
    if (input[i] === "|") {
      let s = "";
      i++;
      while (i < input.length && input[i] !== "|") {
        s += input[i];
        i++;
      }
      i++;
      tokens.push(s);
      continue;
    }
    let s = "";
    while (i < input.length && !/[\s();]/.test(input[i])) {
      if (input[i] === "|") break;
      s += input[i];
      i++;
    }
    if (s) tokens.push(s);
  }
  return tokens;
}

export function parseSExprs(tokens: string[]): SExpr[] {
  const result: SExpr[] = [];
  let pos = 0;

  function parse(): SExpr {
    if (tokens[pos] === "(") {
      pos++;
      const list: SExpr[] = [];
      while (pos < tokens.length && tokens[pos] !== ")") {
        list.push(parse());
      }
      pos++;
      return list;
    } else {
      return tokens[pos++];
    }
  }

  while (pos < tokens.length) {
    result.push(parse());
  }
  return result;
}

export function parseTerm(sexpr: SExpr, funNames: Set<string>): Term {
  if (typeof sexpr === "string") {
    if (funNames.has(sexpr)) {
      return { type: "app", fun: sexpr, args: [] };
    }
    return { type: "var", name: sexpr };
  }
  const fun = sexpr[0] as string;
  const args = sexpr.slice(1).map((s) => parseTerm(s, funNames));
  return { type: "app", fun, args };
}

export function parseARI(input: string): TRS {
  const tokens = tokenize(input);
  const sexprs = parseSExprs(tokens);

  const funs: { name: string; arity: number }[] = [];
  const funNames = new Set<string>();
  const rules: { lhs: Term; rhs: Term }[] = [];

  for (const sexpr of sexprs) {
    if (!Array.isArray(sexpr)) continue;
    const head = sexpr[0];
    if (head === "format") continue;
    if (head === "fun") {
      const name = sexpr[1] as string;
      const arity = parseInt(sexpr[2] as string, 10);
      funs.push({ name, arity });
      funNames.add(name);
    }
    if (head === "rule") {
      const lhs = parseTerm(sexpr[1], funNames);
      const rhs = parseTerm(sexpr[2], funNames);
      rules.push({ lhs, rhs });
    }
  }

  return { funs, rules };
}

File: plugins/trs/src/parse.rs
use std::collections::HashSet;

use crate::types::{SExpr, Term, TRS};

/// Tokenize ARI input. Handles:
/// - Whitespace skipping
/// - `;` line comments
/// - `(` and `)` as tokens
/// - `|quoted|` pipe-delimited symbols (content between pipes, without the pipes)
/// - Plain atoms (stop at whitespace, `(`, `)`, `;`, `|`)
pub fn tokenize(input: &str) -> Vec<String> {
    let chars: Vec<char> = input.chars().collect();
    let mut tokens = Vec::new();
    let mut i = 0;

    while i < chars.len() {
        let c = chars[i];

        // Skip whitespace
        if c.is_whitespace() {
            i += 1;
            continue;
        }

        // Line comment
        if c == ';' {
            while i < chars.len() && chars[i] != '\n' {
                i += 1;
            }
            continue;
        }

        // Parentheses
        if c == '(' || c == ')' {
            tokens.push(c.to_string());
            i += 1;
            continue;
        }

        // Pipe-quoted symbol: |...|
        if c == '|' {
            i += 1;
            let mut s = String::new();
            while i < chars.len() && chars[i] != '|' {
                s.push(chars[i]);
                i += 1;
            }
            i += 1; // consume closing |
            tokens.push(s);
            continue;
        }

        // Plain atom: stop at whitespace, (, ), ;, |
        let mut s = String::new();
        while i < chars.len() {
            let ch = chars[i];
            if ch.is_whitespace() || ch == '(' || ch == ')' || ch == ';' || ch == '|' {
                break;
            }
            s.push(ch);
            i += 1;
        }
        if !s.is_empty() {
            tokens.push(s);
        }
    }

    tokens
}

/// Parse a flat token list into a list of S-expressions.
pub fn parse_sexprs(tokens: &[String]) -> Vec<SExpr> {
    let mut result = Vec::new();
    let mut pos = 0;
    while pos < tokens.len() {
        let (sexpr, new_pos) = parse_one(tokens, pos);
        pos = new_pos;
        result.push(sexpr);
    }
    result
}

fn parse_one(tokens: &[String], pos: usize) -> (SExpr, usize) {
    if tokens[pos] == "(" {
        let mut pos = pos + 1;
        let mut list = Vec::new();
        while pos < tokens.len() && tokens[pos] != ")" {
            let (child, new_pos) = parse_one(tokens, pos);
            pos = new_pos;
            list.push(child);
        }
        // consume the closing ")"
        (SExpr::List(list), pos + 1)
    } else {
        (SExpr::Atom(tokens[pos].clone()), pos + 1)
    }
}

/// Parse a term from an S-expression, given the set of known function names.
pub fn parse_term(sexpr: &SExpr, fun_names: &HashSet<String>) -> Term {
    match sexpr {
        SExpr::Atom(s) => {
            if fun_names.contains(s) {
                Term::App {
                    fun: s.clone(),
                    args: vec![],
                }
            } else {
                Term::Var(s.clone())
            }
        }
        SExpr::List(items) => {
            let fun = match &items[0] {
                SExpr::Atom(s) => s.clone(),
                _ => panic!("expected atom as function head"),
            };
            let args = items[1..].iter().map(|s| parse_term(s, fun_names)).collect();
            Term::App { fun, args }
        }
    }
}

/// Parse an ARI-format string into a TRS.
pub fn parse_ari(input: &str) -> TRS {
    let tokens = tokenize(input);
    let sexprs = parse_sexprs(&tokens);

    let mut funs = Vec::new();
    let mut fun_names = HashSet::new();
    let mut rules = Vec::new();

    for sexpr in &sexprs {
        match sexpr {
            SExpr::List(items) if !items.is_empty() => {
                let head = match &items[0] {
                    SExpr::Atom(s) => s.as_str(),
                    _ => continue,
                };
                match head {
                    "format" => {}
                    "fun" => {
                        if items.len() >= 3 {
                            let name = match &items[1] {
                                SExpr::Atom(s) => s.clone(),
                                _ => continue,
                            };
                            let arity: usize = match &items[2] {
                                SExpr::Atom(s) => s.parse().unwrap_or(0),
                                _ => continue,
                            };
                            fun_names.insert(name.clone());
                            funs.push(crate::types::FunDecl { name, arity });
                        }
                    }
                    "rule" => {
                        if items.len() >= 3 {
                            let lhs = parse_term(&items[1], &fun_names);
                            let rhs = parse_term(&items[2], &fun_names);
                            rules.push(crate::types::Rule { lhs, rhs });
                        }
                    }
                    _ => {}
                }
            }
            _ => {}
        }
    }

    TRS { funs, rules }
}
